File: C#/Elearning3.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Channels;
using System.Threading.Tasks;

Console.OutputEncoding = System.Text.Encoding.UTF8;

await lapMang();

async Task lapMang()
{
    while (true)
    {
        Console.WriteLine("\n--- Mo phong cac ky thuat bat dong bo trong C# ---");
        Console.WriteLine("1) Async/await (mo phong tac vu I/O)");
        Console.WriteLine("2) Parallel.For (tac vu CPU)");
        Console.WriteLine("3) Tasks + Task.WhenAll (chay dong thoi)");
        Console.WriteLine("4) Producer - Consumer (BlockingCollection)");
        Console.WriteLine("5) Channel + IAsyncEnumerable (dong du lieu bat dong bo)");
        Console.WriteLine("6) Gioi han tac vu bang SemaphoreSlim");
        Console.WriteLine("7) Huy tac vu voi CancellationToken");
        Console.WriteLine("0) Thoat");
        Console.Write("Lua chon: ");
        var key = Console.ReadLine();
        Console.WriteLine();

        switch (key)
        {
            case "1": await demoAsyncAwait(); break;
            case "2": demoParallelFor(); break;
            case "3": await demoTasksWhenAll(); break;
            case "4": demoProducerConsumer(); break;
            case "5": await demoChannelAsync(); break;
            case "6": await demoSemaphore(); break;
            case "7": await demoCancellation(); break;
            case "0": return;
            default: Console.WriteLine("Lua chon khong hop le"); break;
        }
    }
}

// 1. Async/Await (mo phong I/O)
async Task demoAsyncAwait()
{
    Console.WriteLine("Demo: async/await — mo phong goi API nhieu lan (I/O-bound)");
    var dsUrl = Enumerable.Range(1, 5).Select(i => $"https://api/{i}").ToArray();

    var dongHo = System.Diagnostics.Stopwatch.StartNew();

    var congViec = dsUrl.Select(u => giaLapTaiDuLieuAsync(u)).ToArray();
    var ketQua = await Task.WhenAll(congViec);

    dongHo.Stop();
    Console.WriteLine($"Hoan thanh {ketQua.Length} tac vu trong {dongHo.ElapsedMilliseconds} ms");
    foreach (var kq in ketQua)
        Console.WriteLine(kq);
}

async Task<string> giaLapTaiDuLieuAsync(string url)
{
    var rd = new Random(Guid.NewGuid().GetHashCode());
    int cho = rd.Next(300, 1000);
    await Task.Delay(cho);
    return $"[OK] {url} (cho {cho} ms)";
}

// 2. Parallel.For (CPU-bound)
void demoParallelFor()
{
    Console.WriteLine("Demo: Parallel.For — chay nhieu tac vu CPU song song");
    int n = 20_000_000;
    Console.WriteLine($"Tinh tong sqrt(i) voi i = 1..{n}");

    var dongHo = System.Diagnostics.Stopwatch.StartNew();
    double tong = 0.0;
    object khoa = new object();

    Parallel.For(1, n + 1, () => 0.0, (i, state, cucBo) =>
    {
        cucBo += Math.Sqrt(i);
        return cucBo;
    }, cucBo =>
    {
        lock (khoa) tong += cucBo;
    });

    dongHo.Stop();
    Console.WriteLine($"Ket qua ~ {tong:N0} (mat {dongHo.ElapsedMilliseconds} ms)");
}

// 3. Task.WhenAll (chay nhieu tac vu)
async Task demoTasksWhenAll()
{
    Console.WriteLine("Demo: Task.WhenAll — tao nhieu tac vu chay dong thoi");
    var dongHo = System.Diagnostics.Stopwatch.StartNew();

    var dsTacVu = new List<Task<int>>();
    for (int i = 0; i < 15; i++)
    {
        int id = i;
        dsTacVu.Add(Task.Run(() => congViecGiaLap(id)));
    }

    int[] kq = await Task.WhenAll(dsTacVu);
    dongHo.Stop();
    Console.WriteLine($"Hoan thanh {kq.Length} tac vu trong {dongHo.ElapsedMilliseconds} ms");
    Console.WriteLine("Ket qua: " + string.Join(", ", kq));
}

int congViecGiaLap(int id)
{
    Thread.SpinWait(150_000);
    Thread.Sleep(50);
    return id * id;
}

// 4. Producer-Consumer (BlockingCollection)
void demoProducerConsumer()
{
    Console.WriteLine("Demo: Producer - Consumer voi BlockingCollection");
    var boDem = new BlockingCollection<int>(10);

    var nguoiTao = Task.Run(() =>
    {
        for (int i = 1; i <= 20; i++)
        {
            boDem.Add(i);
            Console.WriteLine($"Tao {i}");
            Thread.Sleep(70);
        }
        boDem.CompleteAdding();
    });

    var nguoiTieuThu = Enumerable.Range(1, 3).Select(id => Task.Run(() =>
    {
        foreach (var so in boDem.GetConsumingEnumerable())
        {
            Console.WriteLine($"  Nguoi {id} xu ly {so}");
            Thread.Sleep(200);
        }
        Console.WriteLine($"  Nguoi {id} xong viec");
    })).ToArray();

    Task.WaitAll(new[] { nguoiTao }.Concat(nguoiTieuThu).ToArray());
    Console.WriteLine("Ket thuc mo phong Producer-Consumer");
}

// 5. Channel + IAsyncEnumerable
async Task demoChannelAsync()
{
    Console.WriteLine("Demo: Channel + IAsyncEnumerable — mo phong luong du lieu bat dong bo");
    var kenh = Channel.CreateBounded<int>(new BoundedChannelOptions(5)
    {
        SingleWriter = true,
        SingleReader = false
    });

    var ghiKenh = Task.Run(async () =>
    {
        for (int i = 1; i <= 15; i++)
        {
            await kenh.Writer.WriteAsync(i);
            Console.WriteLine($"[Ghi] {i}");
            await Task.Delay(100);
        }
        kenh.Writer.Complete();
    });

    await foreach (var v in docDuLieuAsync(kenh.Reader))
    {
        Console.WriteLine($"[Doc] {v}");
        await Task.Delay(250);
    }

    await ghiKenh;
    Console.WriteLine("Ket thuc mo phong Channel");
}

async IAsyncEnumerable<int> docDuLieuAsync(ChannelReader<int> doc)
{
    while (await doc.WaitToReadAsync())
    {
        while (doc.TryRead(out var muc))
            yield return muc;
    }
}

// 6. SemaphoreSlim
async Task demoSemaphore()
{
    Console.WriteLine("Demo: SemaphoreSlim — gioi han so luong tac vu chay dong thoi");
    var semaphore = new SemaphoreSlim(3);
    var ds = new List<Task>();
    for (int i = 1; i <= 10; i++)
    {
        int id = i;
        await semaphore.WaitAsync();
        ds.Add(Task.Run(async () =>
        {
            try
            {
                Console.WriteLine($"Tac vu {id} bat dau");
                await Task.Delay(700 + id * 50);
                Console.WriteLine($"Tac vu {id} hoan thanh");
            }
            finally
            {
                semaphore.Release();
            }
        }));
    }
    await Task.WhenAll(ds);
    Console.WriteLine("Tat ca tac vu da hoan thanh");
}

// 7. CancellationToken
async Task demoCancellation()
{
    Console.WriteLine("Demo: Huy tac vu bang CancellationToken");
    using var cts = new CancellationTokenSource();

    var tacVuDai = tacVuDaiHanAsync(cts.Token);

    Console.WriteLine("Nhan 'c' de huy trong 5 giay...");
    var phim = Task.Run(() =>
    {
        var k = Console.ReadKey(intercept: true);
        if (k.KeyChar == 'c' || k.KeyChar == 'C')
            cts.Cancel();
    });

    cts.CancelAfter(TimeSpan.FromSeconds(5));

    try
    {
        await tacVuDai;
        Console.WriteLine("Tac vu da hoan thanh binh thuong");
    }
    catch (OperationCanceledException)
    {
        Console.WriteLine("Tac vu da bi huy");
    }
}

async Task tacVuDaiHanAsync(CancellationToken token)
{
    for (int i = 1; i <= 20; i++)
    {
        token.ThrowIfCancellationRequested();
        Console.WriteLine($"Dang xu ly buoc {i}");
        await Task.Delay(700, token);
    }
}

File: C++/Elearning3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <thread>
#include <future>
#include <mutex>
#include <condition_variable>
#include <chrono>
#include <queue>
#include <atomic>
#include <algorithm>
#include <bits/stdc++.h>
using namespace std;



// -------------------- 1. Async/Await (I/O) --------------------
string giaLapTaiDuLieu(const string& url) {
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> dist(300, 1000);
    int delay = dist(gen);
    this_thread::sleep_for(chrono::milliseconds(delay));
    return "[OK] " + url + " (chờ " + to_string(delay) + " ms)";
}

void demoAsyncAwait() {
    cout << "Demo 1: async/await — mô phỏng tải dữ liệu I/O\n";
    vector<string> dsUrl;
    for (int i = 1; i <= 5; ++i)
        dsUrl.push_back("https://api/" + to_string(i));

    auto start = chrono::high_resolution_clock::now();
    vector<future<string>> futures;
    for (auto& u : dsUrl)
        futures.push_back(async(launch::async, giaLapTaiDuLieu, u));

    for (auto& f : futures)
        cout << f.get() << endl;

    auto end = chrono::high_resolution_clock::now();
    cout << "Hoàn thành trong "
         << chrono::duration_cast<chrono::milliseconds>(end - start).count()
         << " ms\n";
}

// -------------------- 2. Parallel.For (CPU) --------------------
void demoParallelFor() {
    cout << "Demo 2: Parallel.For — chạy nhiều tác vụ CPU song song\n";
    const int n = 20'000'000;
    double tong = 0.0;
    mutex mtx;

    auto start = chrono::high_resolution_clock::now();
    int numThreads = thread::hardware_concurrency();
    if (numThreads == 0) numThreads = 4;
    int chunk = n / numThreads;

    vector<thread> threads;
    for (int t = 0; t < numThreads; ++t) {
        int startIdx = t * chunk + 1;
        int endIdx = (t == numThreads - 1) ? n : startIdx + chunk - 1;
        threads.emplace_back([&, startIdx, endIdx]() {
            double localSum = 0;
            for (int i = startIdx; i <= endIdx; ++i)
                localSum += sqrt(i);
            lock_guard<mutex> lock(mtx);
            tong += localSum;
        });
    }
    for (auto& th : threads) th.join();

    auto end = chrono::high_resolution_clock::now();
    cout << "Kết quả ~ " << tong << " (mất "
         << chrono::duration_cast<chrono::milliseconds>(end - start).count()
         << " ms)\n";
}

// -------------------- 3. Task.WhenAll --------------------
int congViecGiaLap(int id) {
    this_thread::sleep_for(chrono::milliseconds(50));
    return id * id;
}

void demoTasksWhenAll() {
    cout << "Demo 3: Task.WhenAll — tạo nhiều tác vụ chạy đồng thời\n";
    auto start = chrono::high_resolution_clock::now();
    vector<future<int>> futures;

    for (int i = 0; i < 15; ++i)
        futures.push_back(async(launch::async, congViecGiaLap, i));

    vector<int> kq;
    for (auto& f : futures) kq.push_back(f.get());

    auto end = chrono::high_resolution_clock::now();
    cout << "Hoàn thành " << kq.size() << " tác vụ trong "
         << chrono::duration_cast<chrono::milliseconds>(end - start).count()
         << " ms\n";
    cout << "Kết quả: ";
    for (auto v : kq) cout << v << " ";
    cout << endl;
}

// -------------------- 4. Producer - Consumer --------------------
class BlockingQueue {
    queue<int> q;
    mutex mtx;
    condition_variable cv;
    bool closed = false;
public:
    void push(int val) {
        unique_lock<mutex> lock(mtx);
        q.push(val);
        cv.notify_one();
    }
    bool pop(int& val) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [&]() { return !q.empty() || closed; });
        if (q.empty()) return false;
        val = q.front();
        q.pop();
        return true;
    }
    void close() {
        unique_lock<mutex> lock(mtx);
        closed = true;
        cv.notify_all();
    }
};

void demoProducerConsumer() {
    cout << "Demo 4: Producer - Consumer với BlockingQueue\n";
    BlockingQueue q;

    auto producer = thread([&]() {
        for (int i = 1; i <= 20; ++i) {
            q.push(i);
            cout << "Tạo " << i << endl;
            this_thread::sleep_for(chrono::milliseconds(70));
        }
        q.close();
    });

    vector<thread> consumers;
    for (int id = 1; id <= 3; ++id) {
        consumers.emplace_back([&, id]() {
            int val;
            while (q.pop(val)) {
                cout << "  Người " << id << " xử lý " << val << endl;
                this_thread::sleep_for(chrono::milliseconds(200));
            }
            cout << "  Người " << id << " xong việc\n";
        });
    }

    producer.join();
    for (auto& c : consumers) c.join();
    cout << "Kết thúc mô phỏng Producer-Consumer\n";
}

// -------------------- 5. Channel bất đồng bộ --------------------
class AsyncChannel {
    queue<int> q;
    mutex mtx;
    condition_variable cv;
    bool done = false;
public:
    void write(int v) {
        unique_lock<mutex> lock(mtx);
        q.push(v);
        cv.notify_all();
    }
    bool read(int& out) {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [&]() { return !q.empty() || done; });
        if (q.empty()) return false;
        out = q.front();
        q.pop();
        return true;
    }
    void complete() {
        unique_lock<mutex> lock(mtx);
        done = true;
        cv.notify_all();
    }
};

void demoChannelAsync() {
    cout << "Demo 5: Channel — luồng dữ liệu bất đồng bộ\n";
    AsyncChannel chan;

    auto writer = thread([&]() {
        for (int i = 1; i <= 15; ++i) {
            chan.write(i);
            cout << "[Ghi] " << i << endl;
            this_thread::sleep_for(chrono::milliseconds(100));
        }
        chan.complete();
    });

    auto reader = thread([&]() {
        int v;
        while (chan.read(v)) {
            cout << "  [Đọc] " << v << endl;
            this_thread::sleep_for(chrono::milliseconds(250));
        }
    });

    writer.join();
    reader.join();
    cout << "Kết thúc mô phỏng Channel\n";
}

// -------------------- 6. Giả lập SemaphoreSlim (C++14) --------------------
class SimpleSemaphore {
    mutex mtx;
    condition_variable cv;
    int permits;
public:
    explicit SimpleSemaphore(int count) : permits(count) {}
    void acquire() {
        unique_lock<mutex> lock(mtx);
        cv.wait(lock, [&]() { return permits > 0; });
        --permits;
    }
    void release() {
        unique_lock<mutex> lock(mtx);
        ++permits;
        cv.notify_one();
    }
};

void demoSemaphore() {
    cout << "Demo 6: SemaphoreSlim (giả lập)\n";
    SimpleSemaphore sem(3);
    vector<thread> tasks;

    for (int i = 1; i <= 10; ++i) {
        tasks.emplace_back([i, &sem]() {
            sem.acquire();
            cout << "Tác vụ " << i << " bắt đầu\n";
            this_thread::sleep_for(chrono::milliseconds(700 + i * 50));
            cout << "Tác vụ " << i << " hoàn thành\n";
            sem.release();
        });
    }

    for (auto& t : tasks) t.join();
    cout << "Tất cả tác vụ đã hoàn thành\n";
}

// -------------------- 7. CancellationToken --------------------
void demoCancellation() {
    cout << "Demo 7: Hủy tác vụ bằng cờ atomic\n";
    atomic<bool> cancel{false};

    thread longTask([&]() {
        for (int i = 1; i <= 20; ++i) {
            if (cancel) {
                cout << "Tác vụ bị hủy tại bước " << i << endl;
                return;
            }
            cout << "Đang xử lý bước " << i << endl;
            this_thread::sleep_for(chrono::milliseconds(700));
        }
        cout << "Tác vụ hoàn thành bình thường.\n";
    });

    cout << "Nhấn Enter để hủy trong 5 giây...\n";
    auto start = chrono::steady_clock::now();
    while (chrono::steady_clock::now() - start < chrono::seconds(5)) {
        if (cin.peek() != EOF) {
            cancel = true;
            break;
        }
        this_thread::sleep_for(chrono::milliseconds(100));
    }

    cancel = true;
    longTask.join();
}

// -------------------- MAIN MENU --------------------
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    while (true) {
        cout << "\n--- Mô phỏng các kỹ thuật bất đồng bộ trong C++ ---\n";
        cout << "1) Async/await (I/O)\n";
        cout << "2) Parallel.For (CPU)\n";
        cout << "3) Task.WhenAll (đồng thời)\n";
        cout << "4) Producer-Consumer\n";
        cout << "5) Channel bất đồng bộ\n";
        cout << "6) SemaphoreSlim (giả lập)\n";
        cout << "7) CancellationToken\n";
        cout << "0) Thoát\n";
        cout << "Lựa chọn: ";

        string key;
        cin >> key;
        cout << endl;

        if (key == "1") demoAsyncAwait();
        else if (key == "2") demoParallelFor();
        else if (key == "3") demoTasksWhenAll();
        else if (key == "4") demoProducerConsumer();
        else if (key == "5") demoChannelAsync();
        else if (key == "6") demoSemaphore();
        else if (key == "7") demoCancellation();
        else if (key == "0") break;
        else cout << "Lựa chọn không hợp lệ!\n";
    }
    return 0;
}
